### ISSUE TITLE
Set offset topic

### ISSUE THREAD

**Issue by boban74** [spring-projects/spring-integration-kafka#146]
Hi ...

I have switched from spring-integration-kafka 1.3 to 2.1 because of apache kafka upgrade to 0.10.x.x. In spring-integration-kafka 1.3 it was able to set offset topic (name) over the OffsetManager. In spring-integration-kafka 2.1 the OffsetManager is not available. How can I set the name of offset topic?

Regards, Boban

**Comment by artembilan** [comment]
Starting with version `2.0` spring-integration-kafka is based on a new foundation - [Spring for Apache Kafka](http://projects.spring.io/spring-kafka/). 
Now you have to specify offset to listen to for the `KafkaMessageListenerContainer` via `TopicPartitionInitialOffset`.

Does it make sense for you?

**Comment by boban74** [comment]
Hi ...
thank you, I'm already using the TopicPartitionInitialOffset but there I can not set the offset topic name. The spring kafka is always creating a topic named __consumer_offsets where the offset of the consumers are stored, this name i would like to set, change. In spring-integration-kafka 1.3 this was possible to do it with the OffsetManager.

**Comment by artembilan** [comment]
I don't understand your explanation.
Please, try to re-phrase it somehow.
```
/**
	 * Construct an instance with the provided initial offset with
	 * {@link #isRelativeToCurrent()} false.
	 * @param topic the topic.
	 * @param partition the partition.
	 * @param initialOffset the initial offset.
	 * @see #TopicPartitionInitialOffset(String, int, Long, boolean)
	 */
	public TopicPartitionInitialOffset(String topic, int partition, Long initialOffset) {
		this(topic, partition, initialOffset, false);
	}
```
And exactly this is used in the:
```
/**
	 * Construct an instance with the supplied configuration properties and specific
	 * topics/partitions/initialOffsets.
	 * @param consumerFactory the consumer factory.
	 * @param containerProperties the container properties.
	 * @param topicPartitions the topics/partitions; duplicates are eliminated.
	 */
	public KafkaMessageListenerContainer(ConsumerFactory<K, V> consumerFactory,
			ContainerProperties containerProperties, TopicPartitionInitialOffset... topicPartitions) {
```
where the last one should be injected into `KafkaMessageDrivenChannelAdapter`.

What am I missing?

Right, there is no an `OffsetManager` abstraction yet, but that doesn't mean that we can't control offset manually and seek it before `KafkaMessageListenerContainer` initialization.

Spring Kafka doesn't create topics. That is Apache Kafka responsibility.

**Comment by boban74** [comment]
Hi,
sorry for my bad English maybe some code will help. I've already done what you've described and it works fine ... I've understand that it is possible to seek offset manually with the TopicPartitionInitialOffset. Here is my code with spring-integration-kafka 1.3.x
```
@Bean
public OffsetManager kafkaOffsetManager() {
    ZookeeperConnect zookeeperConnect = new ZookeeperConnect(zookeeperAddress);
        
    return new KafkaTopicOffsetManager(zookeeperConnect, "ConsumerOffsetTopic");
}

@Bean
public KafkaMessageListenerContainer chatCollectorKafkaMessageListenerContainer() {
    KafkaMessageListenerContainer kafkaMessageListenerContainer = new KafkaMessageListenerContainer(
            kafkaBrokerConnectionFactory(),
            new Partition("ConsumeTopic", 0));
    
    kafkaMessageListenerContainer.setConcurrency(1);
    kafkaMessageListenerContainer.setOffsetManager(kafkaOffsetManager());

    return kafkaMessageListenerContainer;
}

@Bean
public KafkaMessageDrivenChannelAdapter chatCollectorKafkaMessageDrivenChannelAdapter() {
    KafkaMessageDrivenChannelAdapter kafkaMessageDrivenChannelAdapter = new KafkaMessageDrivenChannelAdapter(
            chatCollectorKafkaMessageListenerContainer());

    kafkaMessageDrivenChannelAdapter.setOutputChannel(chatCollectorReceivedChannel());
    kafkaMessageDrivenChannelAdapter.setAutoStartup(true);

    return kafkaMessageDrivenChannelAdapter;
}
```
In KafkaTopicOffsetManager is a topic name "ConsumerOffsetTopic" set and this topic is created by spring integration. In this topic the offset of my topic "ConsumeTopic", which is set in KafkaMessageListenerContainer, is stored.

Here is the code with spring-integration-kafka 2.1.0
```
@Bean
public KafkaMessageListenerContainer<String, String> chatCollectorKafkaMessageListenerContainer() {
    ContainerProperties containerProperties = new ContainerProperties(new TopicPartitionInitialOffset(
            "ConsumeTopic", 0));
    
    KafkaMessageListenerContainer<String, String> kafkaMessageListenerContainer =
            new KafkaMessageListenerContainer<>(consumerFactory(), containerProperties);

    return kafkaMessageListenerContainer;
}

@Bean
public KafkaMessageDrivenChannelAdapter<String, String> chatCollectorKafkaMessageDrivenChannelAdapter() {
    KafkaMessageDrivenChannelAdapter<String, String> kafkaMessageDrivenChannelAdapter =
            new KafkaMessageDrivenChannelAdapter<>(chatCollectorKafkaMessageListenerContainer());

    kafkaMessageDrivenChannelAdapter.setOutputChannel(chatCollectorReceivedChannel());
    kafkaMessageDrivenChannelAdapter.setAutoStartup(true);

    return kafkaMessageDrivenChannelAdapter;
}
```
In both implementations I'm consuming from topic "ConsumeTopic" but in 1.3.x version it was possible to set a topic "ConsumerOffsetTopic" in OffsetManager where the metadata of my consumer are stored. Is it also possible to do it in spring-integration-kafka 2.1?

**Comment by artembilan** [comment]
>  I've understand that it is possible to seek offset manually with the TopicPartitionInitialOffset.

Correct, `TopicPartitionInitialOffset` is exactly for manual offset assignments. Exactly for what the `OffsetManager` has been used before.

> In KafkaTopicOffsetManager is a topic name "ConsumerOffsetTopic" set and this topic is created by spring integration.

No it can't be created. And that doesn't sound reasonable. If you want to seek from some other position then start, that really means that your topic is already there and has some data to skip with desired logic.

>  OffsetManager where the metadata of my consumer are stored. 

No, there is no `OffsetManager` abstraction in Spring Kafka and therefore Spring Integration Kafka 2.x. 
If you need that logic, you should care about that manually.

Follow this issue if that: https://github.com/spring-projects/spring-kafka/issues/9
